feat(curriculum): add Sudoku solver project Python (#52452)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-classes-and-objects-by-building-a-sudoku-solver/6569fca3cd7a9f52f322a298.md

---
id: 6569fca3cd7a9f52f322a298
title: Step 74
challengeType: 20
dashedName: step-74
---

# --description--

> Print the current state of the sudoku board using `print(gameboard) `

Reviewer: ```suggestion
Now, add another `print()` call passing `gameboard` as the argument to print the current state of the sudoku board.
```

# --hints--

> Use the mentioned `print` function to print the current state of the sudoku board.

Reviewer: ```suggestion
You should add another `print()` call to print `gameboard`.
```

```js

> assert.match(code, /print\(gameboard\)/);

Reviewer: ```suggestion
({ test: () => assert.match(e.code.original["main.py"], /(?<=print\('\\nPuzzle to solve:'\).*)^\s{4}print\s*\(\s*gameboard\s*\)/ms) })
```

            for row in range(9)
        )

    def valid_in_square(self, row, col, num):
        row_start = (row // 3) * 3
        col_start=(col // 3) * 3
        for row_no in range(row_start, row_start + 3):
            for col_no in range(col_start, col_start + 3):
                if self.board[row_no][col_no] == num:
                    return False
        return True

    def is_valid(self, empty, num):
        row, col = empty
        valid_in_row = self.valid_in_row(row, num)
        valid_in_col = self.valid_in_col(col, num)
        valid_in_square = self.valid_in_square(row, col, num)
        return all([valid_in_row, valid_in_col, valid_in_square])

    def solver(self):
        if (next_empty := self.find_empty_cell()) is None:
            return True
        else:
            for guess in range(1, 10):
                if self.is_valid(next_empty, guess):
                    row, col = next_empty
                    self.board[row][col] = guess
                    if self.solver():
                        return True
                    self.board[row][col] = 0

        return False

def solve_sudoku(board):
    gameboard = Board(board)
--fcc-editable-region--

>     print('\nPuzzle to solve:')

Reviewer: `print('\nPuzzle to solve:')` requires an additional step.

Author: I have left that intentionally in the seed code.
